src/common/config.h

```diff
@@ -2,84 +2,161 @@
+      key_regex_escaped_(R"rx(^(["'])([^"'=\r\n]+)\1[\t ]*=)rx"),
+      value_regex_(R"rx(^([^#"'=\r\n\t ]+)[\t ]*(?:#.*){0,1}$)rx"),
+      value_regex_escaped_(R"rx(^(["'])([^"'=\r\n]+)\1[\t ]*(?:#.*){0,1}$)rx") {
+    NormalizeEOL(path);
+    fi_.open(path);
+    CHECK(!fi_.fail()) << "Cannot open file " << path;
   }
 
   /*!
-   * \brief parse the configure file
+   * \brief Parse configuration file into key-value pairs.
+   * \param path path to configuration file
+   * \return list of key-value pairs
    */
-  std::vector<std::pair<std::string, std::string> > Parse() {
-    std::vector<std::pair<std::string, std::string> > results{};
+  std::vector<std::pair<std::string, std::string>> Parse() {
+    std::vector<std::pair<std::string, std::string>> results;
     char delimiter = '=';
     char comment = '#';
-    std::string line{};
-    std::string name{};
-    std::string value{};
-
-    while (!fi_.eof()) {
-      std::getline(fi_, line);  // read a line of configure file
-      line = line.substr(0, line.find(comment));  // anything beyond # is comment
-      size_t delimiterPos = line.find(delimiter);  // find the = sign
-      name = line.substr(0, delimiterPos);  // anything before = is the name
-      // after this = is the value
-      value = line.substr(delimiterPos + 1, line.length() - delimiterPos - 1);
-
-      if (line.empty() || name.empty() || value.empty())
-        continue;  // skip a line if # at beginning or there is no value or no name.
-      CleanString(&name);  // clean the string
-      CleanString(&value);
-      results.emplace_back(name, value);
+    std::string line;
+    std::string key, value;
+    // Loop over every line of the configuration file
+    while (std::getline(fi_, line)) {
+      if (ParseKeyValuePair(line, &key, &value)) {
+        results.emplace_back(key, value);
+      }
     }
     return results;
   }
 
-  ~ConfigParse() {
-    fi_.close();
-  }
-
  private:
   std::ifstream fi_;
-  std::string allowableChar_ =
-      "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_-./\\";
+  const std::regex line_comment_regex_, key_regex_, key_regex_escaped_,
+    value_regex_, value_regex_escaped_;
+
+ public:
+  /*!
+   * \brief Normalize end-of-line in a file so that it uses LF for all line
+   *        endings.
+   * This is needed because some OSes use CR or CR LF instead.
+   * So we replace all CR with LF. Replacement will be performed in-place.
+   * \param path path to configuration file
+   */
+  static void NormalizeEOL(const std::string& path) {
+    std::FILE* fp = std::fopen(path.c_str(), "rb+");
+    CHECK(fp) << "Cannot open file " << path;
+    int ch;
+    while ((ch = std::fgetc(fp)) != EOF) {
+      if (ch == static_cast<int>('\r')) {
+        CHECK_EQ(std::fseek(fp, -1, SEEK_CUR), 0)
+          << "Could not normalize line ending in file " << path;
+        std::fputc(static_cast<int>('\n'), fp);
+        CHECK_EQ(std::fseek(fp, 0, SEEK_CUR), 0)
```

Comment:
what is the purpose of this fseek? doesn't the previous fputc advance the file pointer anyways?

Reply:
This line is supposed to be a no-op. We need `std::fseek` here because the C standard requires it when `fgetc()` and `fputc()` are interleaved ([link](https://en.cppreference.com/w/c/io/fopen)):
> In update mode ('+'), both input and output may be performed, but output cannot be followed by input without an intervening call to fflush, fseek, fsetpos or rewind, and input cannot be followed by output without an intervening call to fseek, fsetpos or rewind, unless the input operation encountered end of file.

Reply:
please document it. it is not quite intuitive when you read it the first time.

also, an fflush may be a better option here - at least, it conveys the impression that the streams are sync'd after a write for the readers to see the new data. an fseek from the current position does not quite convey the intent.